[Kang Su Min] iP

kkangs0226 opened the pull request.

Commit:
Add Gradle support

kkangs0226 committed:
Add Level 1 features greet, echo, exit

There is a basic greeting feature but no echo and exit features.

A more personlized chat bot with a unique logo and tone will be more fun.

Let's
* add echo and exit features
* change the logo and tone of the chat bot

kkangs0226 committed:
Add Level 2 features add, list

The chat bot can only echo commands, is unable to store user input.

Adding add and list feature would allow chat bot to store user input
and display past input of user on demand, engaging user.

Let's
* implement add and list feature

kkangs0226 committed:
Add Level 3 feature 'mark as done'

User can only add tasks but cannot mark a task as done.

Adding this future allows users to give 'done' command which
makes chat bot more useful to users when keeping track of their
tasks.

Let's
* add 'mark as done' feature
* add error catching methods to show error message when done command
  is executed with out of bounds index or a non-integer

kkangs0226 committed:
Add Level 4 features todos, events, dealines

User can only add one type of task, and cannot differentiate between
different types of tasks in their list.

Adding child classes to Task - Todo, Event, Deadline would help user
add different types of tasks to their list. Printing the number of
tasks in the user's list after addition of every task will also improve
user experience.

Let's
* implement child classes of Task - Todo, Event, Deadline
* add a method to print the  number of tasks in the message after adding
  every task to the list

kkangs0226 committed:
Add Automated Text UI Testing

Chat bot has no automated testing function and requires testing through
individual commands.

Adding automated testing function would allow faster and more efficient
testing of the chat bot.

Let's
* add automated text UI testing to the program using I/O redirection
technique

kkangs0226 committed:
Add Level 5 feature error handling

Chat bot is unable to handle invalid user input and throws exception,
which affects user experience.

Implementing a DukeException to handle invalid input would improve user
experience as customisation of error messages is possible.

Let's
* create a DukeException class to deal with invalid input
* handle invalid done command with out of bounds index
* handle invalid task command
* handle empty todo command

kkangs0226 committed:
Add Level 6 feature delete

Chat bot only allows for adding tasks but does not support deletion
of tasks.

Adding a delete feature to the chat bot would make it more useful
to users.

Let's
* add a new delete method to handle delete command
* add a new delete message to print to user when a task is deleted

kkangs0226 committed:
Implement enums for commands

Code in main Duke class is long because of multiple boolean methods to check
type of command.

Using enums will shorten code and make it more readable.

Let's
* implement InputCommands enum to represent command type

kkangs0226 committed:
Add Level 7 save feature

Any changes made to the list of tasks does not get saved, and every run
of the program would give an empty list at the start, which is not
useful to the user.

Adding save feature would allow the program to save the changes to
the list of tasks from the previous run of the program.

Let's
* save the list of tasks in a duke.txt file under data folder
  every time there are changes to the list of tasks
* read data from the duke.txt file at the start of the program
  run so that the data from previous program can be called and
  used in the program

kkangs0226 committed:
Add Level 8 features dates and times

The chat bot accepts the date/time as a string input from user,
which means meaningful storage of  dates is not possible as there
is no conversion of the string input to date object.

Importing LocalDate class and storing string input by users as a
Date object will help chat bot understand date better and do more
meaningful operations with date.

Let's
* import LocalDate class and store deadline/event dates as a
  LocalDate object in task objects
* implement new command that prints deadlines/events occurring on
  the date given as string input by user

kkangs0226 committed:
Merge branch 'branch-Level-7'

The changes to the code are in branch-Level-7

Merging branch-Level-7 to master branch will update the code in
master branch to the one in branch-Level-7 and join them back
into one master branch.

Let's
* merge branch-Level-7 to master branch

kkangs0226 committed:
Merge branch 'branch-Level-8'

There are merge conflicts between branch-Level-8 and master branch.

Resolving conflicts and merging the two branches would update my
code in master branch.

Let's
* resolve conflicts and merge branch-Level-8 to master branch

kkangs0226 committed:
Add MoreOOP feature

The methods are not abstracted enough.

Adding more OOP concepts into my code by abstracting classes and
methods will improve readability.

Let's
* create Parser, Storage, TaskList, Ui classes to abstract code
* create *Command classes that inherit from abstract Command class

kkangs0226 committed:
Add A-Packages

Projet code is unorganised because it is not packaged properly.

Packaging code into different packages will make the code more
organised.

Let's
* create package duke
* create sub-packages task and command to group different files
  together

kkangs0226 committed:
Add A-JUnit

Code does not have any JUnit tests.

Adding JUnit tests would improve efficiency when finding
errors in code.

Let's
* add some JUnit tests to test creation of file strings for
  Todo, Event and Deadline tasks.
* add JUnit tests to test whether DukeException is thrown for
  invalid add commands

kkangs0226 committed:
Add JavaDocs to code

Code now has no description to explain what the methods and
classes do.

Adding JavaDocs would help readers understand code better.

Let's
* add JavaDocs to public classes and methods

kkangs0226 committed:
Add A-CodingStandard

Parts of code do not follow coding standards.

Adding coding standards would increase readability of code.

Let’s
* add coding standards to code

kkangs0226 committed:
Add Level-9 find feature

Code does not contain class and methods for keyword search
function.

Adding find feature would allow users to find tasks more
efficiently.

Let's
* add create FindCommand class to represent find command
  and add methods accordingly.

kkangs0226 committed:
Merge branch 'branch-A-JavaDoc'

Branch for JavaDoc is not yet merged into master branch.

Let's
* merge branch-A-JavaDoc into master branch to include JavaDoc in
  master branch

kkangs0226 committed:
Merge branch 'branch-A-CodingStandard'

Branch for coding standard not yet merged into master branch.

Let's
* merge branch-A-CodingStandard into master branch

kkangs0226 committed:
Merge branch 'branch-Level-9'

Branch-Level-9 find feature not yet merged into master branch.

Let's
* merge branch-Level-9 into master branch

florenciamartina commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,126 @@
+
+    List<Task> tasks = new ArrayList<>();
+    String filePath;
+    File dataFile;
```
I think it is better to add access modifier to your variables (private, preferably) to comply with coding standard (and encapsulation) 😄

florenciamartina commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,51 @@
+    public static void main(String[] args) {
+        new Duke("data/duke.txt").run();
+    }
+}
```
It's missing the newline at the end of the file. You might want to consider doing checkstyle in IntelliJ as it highlights all coding style violation! (makes our life easier) 😄

florenciamartina commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,19 @@
+ */
+public class DukeException extends Exception {
+
+    String message;
```
In my opinion, you don't need the message attribute since Exception has it.

florenciamartina commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,19 @@
+        message = input;
+    }
+
+    @Override
+    public String toString() {
+        return "Apple Pineapple!! " + message;
+    }
```
Following the previous comment, there is no need to write toString() method since Exception class already has a built-in function getMessage() to fetch the error message.

florenciamartina commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,134 @@
+package duke;
+
+import duke.command.Command;
+import duke.command.ExitCommand;
+import duke.command.AddCommand;
+import duke.command.DoneCommand;
+import duke.command.DeleteCommand;
+import duke.command.InputCommand;
+import duke.command.GetCommand;
+import duke.command.ListCommand;
+import duke.command.FindCommand;
```
It is good that you group the import statements consistently. However (this is just nitpicking, not compulsory), it looks neater if you write them in lexicographical order.

florenciamartina commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,134 @@
+            return new DoneCommand(input);
+        case DELETE:
+            return new DeleteCommand(input);
+        case TODO:
+        case DEADLINE:
```
May I ask why do you leave these cases empty? Is it because Todo, Deadline, and Event are considered as AddCommand? Perhaps you might want to consider separating them (TodoCommand, DeadlineCommand, EventCommand extending from AddCommand) to make it more OOP.

florenciamartina commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,134 @@
+     * @return true if valid, false if invalid.
+     */
+    public static boolean isValidIndex(String input, int tasksNumber) {
+        String[] stringArray = input.split(" ");
```
Perhaps you might want to consider renaming this variable to something like "parsedInput" or "splittedInput" to avoid confusion? (stringArray is a bit too general) 😆 I noticed the same issue in several other places too.

florenciamartina commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,134 @@
+     * @return Index in input.
+     */
+    public static int getIndex(String input) {
+        String[] stringArray = input.split(" ");
```
To avoid repetition, you might want to consider making the parsedInput as class variable. So you don't have to parsed multiple times.

florenciamartina commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,134 @@
+     */
+    public static String getTodoDescription(String input) throws DukeException{
+        try {
+            return input.substring(5);
```
I think it is better to avoid magic number because it is confusing. Just like my previous comment, I think it is better to store the parsedInput to make debugging easier! I noticed this same issue in several other places too, you might want to consider this! 😄

florenciamartina commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,134 @@
+        }
+    }
+
+    /**
+     * Gets array of deadline description strings.
+     * @param input String input of user.
+     * @return Array of deadline description strings.
+     */
+    public static String[] getDeadlineStrings(String input) {
+        String[] stringArray = input.split(" /by ");
+        stringArray[0] = stringArray[0].substring(9);
+        return stringArray;
+    }
+
+    /**
+     * Gets array of event description strings.
+     * @param input String input of user.
+     * @return Array of event description strings.
+     */
+    public static String[] getEventTimeStrings(String input) {
+        String[] stringArray = input.split(" /at ");
+        stringArray[0] = stringArray[0].substring(6);
+        return stringArray;
+    }
```
I think it's really great that you abstract this to methods! It makes debugging easier (you might want to avoid magic number though!)

florenciamartina commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,126 @@
+    public List<Task> load() throws DukeException{
+
+        Path directoryPath = Paths.get("data");
+        boolean directoryExists = Files.exists(directoryPath);
```
To make it sounds more like boolean, you might want to rename this to "hasDirectory" instead.

florenciamartina commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,126 @@
+                }
+                return tasks;
+            } catch (IOException e) {
+                throw new DukeException("WHAT WHATS GOING ON");
```
I think this part is great since it is easy to read. Just some nitpicking, though, I think it is better if your error message is more specific so the user can know the problem! 👍

florenciamartina commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,126 @@
+     * @param tasks list of objects to add the new task objects to.
+     * @param line Line to be converted to a task object.
+     */
+    private void convertLineToTasks(List<Task> tasks, String line) {
```
Perhaps it is better to rename the method to "readDataToList"?

florenciamartina commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,144 @@
+ */
+public class Ui {
+
+    public static String divider = "______________________________________________________________________________";
+    public static String logo = "             *\n"
+            + "      o  o  / \\  o  o\n"
+            + "      |\\/ \\/   \\/ \\/|\n"
+            + "      |             |\n"
+            + "      |ooooooooooooo|\n"
+            + " __  _  ____  ____    ____      ____    ___   ____\n"
+            + "|  |/ ||    ||    \\  /    |    |    \\  /   \\ |    \\\n"
+            + "|  ' /  |  | |  _  ||   __|    |  o  )|     ||  o  )\n"
+            + "|    \\  |  | |  |  ||  |  |    |     ||  O  ||     |\n"
+            + "|     | |  | |  |  ||  |_ |    |  O  ||     ||  O  |\n"
+            + "|  .  | |  | |  |  ||     |    |     ||     ||     |\n"
+            + "|__|\\_||____||__|__||___,_|    |_____| \\___/ |_____|\n";
```
Assuming these are constants (since they won't change), I think the variable names must be all uppercase (e. g. DIVIDER and LOGO)

florenciamartina commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,73 @@
+    private InputCommand inputCommand;
+
+    public AddCommand(String input, InputCommand inputCommand) {
+        this.exit = false;
```
To make this variable sounds more boolean, you might want to consider renaming this to isExit.

florenciamartina commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,73 @@
+        case TODO:
+            try {
+                taskList.addList(new Todo(Parser.getTodoDescription(input)));
+                ui.printAddedMessage(new Todo(Parser.getTodoDescription(input)), taskList.getListSize());
```
I see that you do taskList.getListSize() several times. I think it is better to store it as a variable listSize to avoid calling getListSize() multiple times.

florenciamartina commented on src/main/java/duke/command/DoneCommand.java:
```diff
@@ -0,0 +1,38 @@
+package duke.command;
+
+import duke.*;
```
It's a good practice to avoid using * when importing. You should only import statements that you are going to use! 😄

florenciamartina commented on src/main/java/duke/command/GetCommand.java:
```diff
@@ -0,0 +1,33 @@
+/**
+ * Represents command that is specific to the get command.
+ */
+public class GetCommand extends Command {
```
May I ask what is GetCommand for? Correct me if I'm wrong but assuming this is for finding task, perhaps you might want to consider renaming this class to FindCommand to avoid confusion. Since get is a bit ambiguous (can also be interpreted as a verb).

florenciamartina commented on src/main/java/duke/command/InputCommand.java:
```diff
@@ -0,0 +1,14 @@
+    DELETE,
+    INVALID,
+    BYE,
+    GET_TASK,
+    FIND,
```
May I ask what is the difference between both of these commands? Maybe you want to rename it something more intuitive (SHOW, FIND_KEYWORD, FIND_DATE, etc) ?

florenciamartina commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,56 @@
+/**
+ * Represents task of user
+ */
+public class Task {
```
You might want to change this class to an abstract class because it doesn't make sense to create a new a Task. It is algo a good practice in OOP because it reduces code repetition.

florenciamartina commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,56 @@
+public class Task {
+    protected String description;
+    protected boolean isDone;
+    String dateString;
+    LocalDate date;
```
Don't forget the access modifier! 😄 It is important as attributes should be declared private when possible and neglecting access modifier makes the attributes implicitly public.

florenciamartina commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,56 @@
+     * Returns string representation of task object to be stored in file
+     * @return string representation of task object for file storage
+     */
+    public String toFileString() {
```
You might want to consider renaming this to comply with the coding standard. Method name should be a verb and intuitive.

florenciamartina reviewed (commented):
Overall, your code is neat and readable. There isn't much problem with the OOP principles either. However, you might want to avoid magic numbers and use more intuitive name. Furthermore, don't forget to write access modifiers when declaring your attributes! Good job on your iP! 

P. S. I like your bot's personality (reminds me of the Minions!😆) Looking forward to the final product.

kkangs0226 committed:
Fix runtest.sh bugs

kkangs0226 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Storage.java
#	src/main/java/duke/Ui.java
#	src/main/java/duke/command/Command.java
#	src/main/java/duke/task/TaskList.java

kkangs0226 committed:
Update runtest.sh

Runtest.sh actual and expected output differs.

Updating expected output file will fix the difference
in output.

Let's
* Update EXPECTED.TXT

kkangs0226 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

kkangs0226 committed:
Add checkstyle configurations

Project now does not have checkstyle configurations.

Adding checkstyle configurations will allow users to run
checkstyle locally.

Let's
* add checkstyle configurations

kkangs0226 committed:
Update code to follow checkstyle.

Code has a few coding style violations.

Updating code to follow checkstyle would make the code follow
the conventional coding style and would make code more readable.

Let's
* use checkstyle on all files to find checkstyle errors
* update code accordingly to follow checkstyle

kkangs0226 committed:
Merge branch 'branch-A-Checkstyle'

branch-A-Checkstyle is not yet merged to the master branch.

Merging would update master branch to follow checkstyle as well.

Let's merge branch 'branch-A-Checkstyle' to master branch.

kkangs0226 committed:
Add GUI to Chat Bot

The Chat bot now is a command line app.

Adding GUI would make it more visually appealing and interactive.

Let's
* add GUI features using javafx

kkangs0226 committed:
Merge branch 'branch-Level-10'

branch-Level-10 not yet merged into master branch.

Merging branch would incorporate GUI features into master branch.

Let's merge branch-Level-10 into master branch.

kkangs0226 committed:
Create jar file

kkangs0226 committed:
Add assertions to program

Program does not have assertions to verify the truth of certain
assumptions.

Adding assertions would help document important assumptions that
should hold at certain points in code.

Let's add assertions to program.

kkangs0226 committed:
Merge pull request #1 from kkangs0226/branch-A-Assertions

Add assertions to program

kkangs0226 committed:
Improve code quality

Code does not comply with some coding standards.

Refactoring code to comply with coding standards
would improve quality of code.

Let's
* check code quality using Java coding conventions
* refactor code where necessary

kkangs0226 committed:
Merge pull request #2 from kkangs0226/branch-A-CodeQuality

Improve code quality

kkangs0226 committed:
Add extension for detecting duplicates

Program unable to detect duplicates.

Let's add methods to detect duplicates.

kkangs0226 committed:
Add streams and lambda to code

Program does not use lambdas and streams.

Using lambdas and streams will shorten code.

Let's add lambdas and streams to program where
applicable.

kkangs0226 committed:
Add A-BetterGui

The GUI of the chat bot can be improved.

Adding design that is in line with the personality of the chat
bot would make user experience more enjoyable.

Let's
* improve on the GUI of the chat bot by adding background, color,
  changing font family and size

kkangs0226 committed:
Add A-UserGuide

Adding User Guide helps users

Let's add User Guide

kkangs0226 committed:
Update User Guide

User Guide has no images.

Adding a screenshot of the chat bot will make User Guide more
readable and enjoyable.

Let's
* add screenshot of chat bot
* add credits to user guide

kkangs0226 committed:
Fix Features header on User Guide

Features header not showing.

Let's try to remove the space to fix the problem.

kkangs0226 committed:
Update User Guide and Ui.png

There are several mistakes in the User Guide. Ui.png also has
lines at the side.

Updating user guide and replacing image would improve user
guide.

Let's
* update user guide
* update Ui.png

kkangs0226 committed:
Fix errors in Storage

Commands unable to run properly when app is opened for the first
time.

Let's fix null pointer error in the Storage class.

kkangs0226 committed:
Fix whitespace in User Guide
